js/src/dht/provide.js

```diff
@@ -0,0 +1,102 @@
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const CID = require('cids')
+const { getDescribe, getIt } = require('../utils/mocha')
+
+const expect = chai.expect
+chai.use(dirtyChai)
+
+module.exports = (createCommon, options) => {
+  const describe = getDescribe(options)
+  const it = getIt(options)
+  const common = createCommon()
+
+  describe('.dht.provide', function () {
+    this.timeout(80 * 1000)
+
+    let ipfs
+
+    before(function (done) {
+      // CI takes longer to instantiate the daemon, so we need to increase the
+      // timeout for the before step
+      this.timeout(60 * 1000)
+
+      common.setup((err, factory) => {
+        expect(err).to.not.exist()
+
+        factory.spawnNode((err, node) => {
+          expect(err).to.not.exist()
+          ipfs = node
+          done()
+        })
+      })
+    })
+
+    after((done) => common.teardown(done))
+
+    it('should provide local CID', (done) => {
+      ipfs.files.add(Buffer.from('test'), (err, res) => {
+        if (err) return done(err)
+
+        ipfs.dht.provide(new CID(res[0].hash), (err) => {
+          expect(err).to.not.exist()
+          done()
+        })
```

Comment:
I don't believe many of these tests will work anymore as there will never be any peers on the dht to provide too, and there isn't much of a reason to provide to `self`.

The dht for go-ipfs I know will error if no peers are connected when running `provides` [0]. Should this be valid, and an issue opened for go-ipfs?



[0] https://github.com/ipfs/go-ipfs/blob/08fb11fa896704d9e081bf64e66fac1f6d9d03dc/core/commands/dht.go#L280

Reply:
I’ve run these tests against js-ipfs and js-ipfs-api (talks to go-ipfs) and they all pass, so the tests that could fail might already be skipped...I’ll check and get back to you

Reply:
DHT tests are completely commented out in js-ipfs currently 😟 https://github.com/ipfs/js-ipfs/blob/master/test/core/interface/dht.js

Some tests in js-ipfs-api were [already skipped](https://github.com/ipfs/js-ipfs-api/pull/785/files#diff-fd1dd60e18a7ade9d795b77bda4aff2eR47), but not any `provide` tests. I will investigate further...

Reply:
These tests are passing for me for js-ipfs-api:

<img width="421" alt="screen shot 2018-06-25 at 11 31 32" src="https://user-images.githubusercontent.com/152863/41845433-5fc1ea3a-786b-11e8-97a9-2b7c8937d2fc.png">

Is it because they are providing to the bootstrap nodes?

Reply:
Historically there were only a couple of DHT tests. The suite was increased with https://github.com/ipfs/interface-ipfs-core/pull/288 which led me to found some issues with the http-api.